### PR TITLE
misc work

### PR DESCRIPTION
**Pull request by est77**
- Fix orientation of tangents in ShadingPoint.
- Fix orientation of tangents in OSL.
- Added APPLESEED_LIKELY/UNLIKELY branch prediction macros for gcc.
- Added Basis3d to python bindings.
- Fallback to the original shading basis if the normal from OSL is zero.
- Small microfacet related code refactors.

### CODE REVIEW

File: src/appleseed.python/bindbasis.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+using namespace std;
+
+namespace
+{
+    template <typename T>
+    void do_bind_basis(const char* class_name)
+    {
+        bpy::class_<Basis3<T> >(class_name)
+            .def(bpy::init<>())
+            .def(bpy::init<Vector<T,3> >())
```
Comment (dictoon) [inline src/appleseed.python/bindbasis.cpp]: `<T, 3>`

File: src/appleseed/renderer/kernel/shading/shadingpoint.cpp [hunk 2]
```diff
@@ -410,6 +410,10 @@ void ShadingPoint::compute_world_space_partial_derivatives() const
             m_dpdu = (dv1 * dp0 - dv0 * dp1) * rcp_det;
             m_dpdv = (du0 * dp1 - du1 * dp0) * rcp_det;
 
+            // Make sure dPdv x dPdu == N
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/shadingpoint.cpp]: Why should dPdv x dPdu == N?

Comment (est77) [inline src/appleseed/renderer/kernel/shading/shadingpoint.cpp]: The comment is not 100% correct, I'll fix it.
The idea is that dPdv x dPdu and dPdy x dPdx point in the same direction as N.
If not, then we flip dPdu or dPdx.

**Comment by dictoon** [comment]
Couple minor things to fix and then it's good to merge!

**Pull request merged**